src/bmson/bms_to_bmson.rs

```diff
@@ -0,0 +1,356 @@
+    #[error("stop duration was invalid, using default value")]
+    InvalidStopDuration,
+    /// The note lane was invalid and default value was used.
+    #[error("note lane was invalid, using default value")]
+    InvalidNoteLane,
+    /// The initial BPM was missing and default value was used.
+    #[error("initial BPM was missing, using default value")]
+    MissingBpm,
+    /// The total percentage was missing and default value was used.
+    #[error("total percentage was missing, using default value")]
+    MissingTotal,
+}
+
+/// Output of the conversion from `Bms` to `Bmson`.
+#[derive(Debug, Clone, PartialEq)]
+pub struct BmsToBmsonOutput {
+    /// The converted `Bmson` object.
+    pub bmson: Bmson,
+    /// Warnings that occurred during the conversion.
+    pub warnings: Vec<BmsToBmsonWarning>,
+}
+
+impl Bms {
+    /// Convert `Bms` to `Bmson`.
+    pub fn to_bmson(self) -> BmsToBmsonOutput {
+        let mut warnings = Vec::new();
+        let converter = PulseConverter::new(&self);
+
+        const EASY_WIDTH: f64 = 21.0;
+        const VERY_EASY_WIDTH: f64 = EASY_WIDTH * 1.25;
+        const NORMAL_WIDTH: f64 = 18.0;
+        const HARD_WIDTH: f64 = 15.0;
+        const VERY_HARD_WIDTH: f64 = 8.0;
+        let judge_rank = FinF64::new(match self.header.rank {
+            Some(JudgeLevel::OtherInt(4)) => VERY_EASY_WIDTH / NORMAL_WIDTH, // VeryEasy implementation of beatoraja.
+            Some(JudgeLevel::Easy) => EASY_WIDTH / NORMAL_WIDTH,
+            Some(JudgeLevel::Normal) | None => 1.0,
+            Some(JudgeLevel::Hard) => HARD_WIDTH / NORMAL_WIDTH,
+            Some(JudgeLevel::VeryHard) => VERY_HARD_WIDTH / NORMAL_WIDTH,
+            Some(JudgeLevel::OtherInt(_)) => 1.0,
+        })
+        .unwrap_or_else(|| {
+            warnings.push(BmsToBmsonWarning::InvalidJudgeRank);
+            FinF64::new(1.0).expect("Internal error: 1.0 is not a valid FinF64")
+        });
+
+        let resolution = self.resolution_for_pulses();
+
+        let last_obj_time = self
+            .last_obj_time()
+            .unwrap_or_else(|| ObjTime::new(0, 0, 4));
+        let lines = (0..=last_obj_time.track.0)
+            .map(|track| BarLine {
+                y: converter.get_pulses_on(Track(track)),
+            })
+            .collect();
+
+        let bpm_events = self
+            .arrangers
+            .bpm_changes
+            .values()
+            .filter_map(|bpm_change| {
+                Some(BpmEvent {
+                    y: converter.get_pulses_at(bpm_change.time),
+                    bpm: FinF64::new(bpm_change.bpm.clone().to_f64().unwrap_or(120.0))
+                        .unwrap_or_else(|| {
+                            warnings.push(BmsToBmsonWarning::InvalidBpm);
+                            FinF64::new(120.0).expect("Internal error: 120.0 is not a valid FinF64")
+                        }),
+                })
+            })
+            .collect();
```

Comment:
This `filter_map` can be replaced with `map` like this:
```suggestion
        let bpm_events =
            self.arrangers
                .bpm_changes
                .values()
                .map(|bpm_change| BpmEvent {
                    y: converter.get_pulses_at(bpm_change.time),
                    bpm: FinF64::new(bpm_change.bpm.clone().to_f64().unwrap_or(120.0))
                        .unwrap_or_else(|| {
                            warnings.push(BmsToBmsonWarning::InvalidBpm);
                            FinF64::new(120.0).expect("Internal error: 120.0 is not a valid FinF64")
                        }),
                })
                .collect();
```